### PR TITLE
Various speed hacks

### PR DESCRIPTION
**Pull request by pitrou**
- minimize the overhead of "safe" allocation
- encourage vectorization by using getelementpointer where possible, rather than pointer addition
- fix a bug in the "is_unit_strided" computation logic in ufuncs
- tell LLVM elements of an array shape are always positive or zero
- allow LLVM to ignore signed overflow on integer arithmetic
- add some attributes to declarations of various external functions (e.g. math or NRT)

Note: I've also added some benchmarks to the benchmark suite at https://github.com/numba/numba-benchmark
(graphs at http://numba.pydata.org/numba-benchmark/ -- probably not up to date)

### CODE REVIEW

**Comment by seibert** [comment]
Is this ready for review?

**Comment by pitrou** [comment]
I was hoping for other use cases to crop up, but yes.

File: numba/cgutils.py [hunk 1]
```diff
@@ -643,19 +651,14 @@ def is_neg_int(builder, val):
...
-            ind = Constant.int(Type.int(32), i)
-        else:
-            ind = i
-        idx.append(ind)
-    return builder.gep(ptr, idx, inbounds=True)
+def gep_inbounds(builder, ptr, *inds, **kws):
+    return gep(builder, ptr, *inds, inbounds=True, **kws)
 
 
 def gep(builder, ptr, *inds, **kws):
```
Comment (hargup) [inline numba/cgutils.py]: @pitrou Can you give a short description of this function? Maybe as a docstring.

Comment (hargup) [inline numba/cgutils.py]: Is `gep` a shorthand for `get element pointer` ?

Comment (pitrou) [inline numba/cgutils.py]: Yes, it is. I'll add some docstrings.

**Comment by seibert** [comment]
Jenkins (I think I'm going to start calling it "The Butler") approves now, so we're ready to merge.

**Pull request merged**